File: credit/credit.c
#include <cs50.h>
#include <stdio.h>
#include <math.h>

int every_other_digit(long credit_card);
int multiplyAndSum(int last_digit);
int number_digits(long credit_card);
bool isValidAmex(long credit_card, int numDigits);
bool isValidMasterCard(long credit_card, int numDigits);
bool isValidVisa(long credit_card, int numDigits);

int main (void)
{
    long credit_card = get_long("Credit Card: ");
    int sum_every_other_digit = every_other_digit(credit_card);
    int numDigits = number_digits(credit_card);
    bool amex = isValidAmex(credit_card, numDigits);
    bool master = isValidMasterCard(credit_card, numDigits);
    bool visa = isValidVisa(credit_card, numDigits);
    if (sum_every_other_digit % 10 != 0){
        printf("INVALID\n");
        return 0;
    }
    else if(amex == true){
        printf("AMEX\n");
    }
    else if(master == true){
        printf("MASTERCARD\n");
    }
    else if(visa == true){
        printf("VISA\n");
    }
    else{
        printf("INVALID\n");
        return 0;
    }
}

int every_other_digit(long credit_card){
    int sum = 0;
    bool isAlternateDigit = false;
    while(credit_card > 0){
        if(isAlternateDigit == true){
            int last_digit = credit_card % 10;
            int product = multiplyAndSum(last_digit);
            sum = sum + product;
        }
        else{
            int last_digit = credit_card % 10;
            sum = sum + last_digit;
        }
        isAlternateDigit = !isAlternateDigit;
        credit_card = credit_card / 10;
    }
    return sum;
}

int multiplyAndSum(int last_digit){
    int multiply = last_digit * 2;
    int sum = 0;
    while(multiply > 0){
        int last_digit_multiply = multiply % 10;
        sum = sum + last_digit_multiply;
        multiply = multiply / 10;
    }
    return sum;
}

int number_digits(long credit_card){
    int count = 0;
    while(credit_card > 0){
        count++;
        credit_card = credit_card / 10;
    }
    return count;
}

bool isValidAmex(long credit_card, int numDigits){
        int i = numDigits - 2;
        int first_two_digits = credit_card / pow (10,i);
        if((numDigits == 15) && (first_two_digits == 34 || first_two_digits == 37)){
            return true;
        }
        else{
            return false;
        }
}

bool isValidMasterCard(long credit_card, int numDigits){
    int i = numDigits - 2;
    int first_two_digits = credit_card / pow (10,i);
    if((numDigits == 16) && (first_two_digits > 50 && first_two_digits < 56)){
        return true;
    }
    else{
        return false;
    }
}

bool isValidVisa(long credit_card, int numDigits){
    if(numDigits == 13 || numDigits == 16){
        int i = numDigits - 1;
        int first_digit = credit_card / pow (10, i);
        if(first_digit == 4){
            return true;
        }
        else{
            return false;
        }
    }
    return false;
}

File: Week6/sentimental-credit/credit.py
from cs50 import get_int

def main():
    credit_card = get_int("Credit Card: ")
    sum_digit = sum_digits(credit_card)
    numDigits = len(str(credit_card))
    amex = isAmex(credit_card, numDigits)
    master = isMaster(credit_card, numDigits)
    visa = isVisa(credit_card, numDigits)
    if sum_digit % 10 != 0:
        print("INVALID")
        exit(0)
    elif amex == True:
        print("AMEX")
    elif master == True:
        print("MASTERCARD")
    elif visa == True:
        print("VISA")
    else:
        print("INVALID")
        exit(0)

def sum_digits(n):
    sum = 0
    AlternateDigit = False
    while n > 0:
        if AlternateDigit == True:
            last_digit = n % 10
            product = multiplyAndSum(last_digit)
            sum += product
        else:
            last_digit = n % 10
            sum += last_digit
        AlternateDigit = not AlternateDigit
        n //= 10
    return sum

def multiplyAndSum(n):
    multiply = n * 2
    sum = 0
    while multiply > 0:
        last_digit = multiply % 10
        sum += last_digit
        multiply //= 10
    return sum

def isAmex(n, p):
    i = p - 2
    first_two_digits = n // 10 ** i
    if p == 15 and first_two_digits == 34 or first_two_digits == 37:
        return True
    else:
        return False

def isMaster(n, p):
    i = p - 2
    first_two_digits = n // 10 ** i
    if p == 16 and 50 < first_two_digits < 56:
        return True
    else:
        return False

def isVisa(n, p):
    if p == 13 or p == 16:
        i = p - 1
        first_digit = n // 10 ** i
        if first_digit == 4:
            return True
        else:
            return False
    return False

main()
